### ISSUE TITLE
Saving an empty file creates an additional new line

### ISSUE THREAD

**Issue by joshuaia** [zed-industries/zed#11044]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

1. Create and open new empty file, for example `vim.txt`
2. Don't type anything, just save the file
3. File will be saved with two new lines inside

Same behaviour in vim mode.

I think there shouldn't be multiple new lines in an empty file.

### Environment

Zed: v0.132.2
OS: macOS Sonoma 14.4.1
Memory: 16GB
Architecture: Apple M2

**Comment by Imgkl** [comment]
I don't see this behaviour on 0.133.2

**Comment by joshuaia** [comment]
Small update, it's important to have `ensure_final_newline_on_save=true`.

**Comment by maxdeviant** [comment]
Just so we're clear, are there two _newlines_ or just two lines with a single newline?

This is what I see:

<img width="433" alt="Screenshot 2024-04-26 at 11 05 32 AM" src="https://github.com/zed-industries/zed/assets/1486634/b63f76a7-576d-431b-8314-fed95caa4a10">

The above behavior is the expected behavior when `ensure_final_newline_on_save` is `true`.

**Comment by joshuaia** [comment]
Yes, your reproduction is correct.
I was expecting no newline without content, for example SublimeText doesn't create a newline in that case, and I think it handles such case better, since it makes no sense to have final newline without content of the file.

By the way, vim works this way too: it doesn't create a final newline with no content.
You can check its behavior this way:
1. Open empty file in Zed and at the same time in vim
2. Save empty file in vim
3. Switch back to Zed and you will see there is no final newline
4. Switch back to Vim and type something
5. Switch back to Zed and you will see final newline
6. Switch back to Vim and remove everything, keep only two empty lines (as on your screenshot)
7. Switch back to Zed and you will see 3 lines in total, since Vim adds final newline by default

**Comment by maxdeviant** [comment]
If you change the `ensure_final_newline_on_save` setting to `false` does that give you the behavior you want?

**Comment by joshuaia** [comment]
Yes, if I'm saving empty file.

Here is how it works now:

https://github.com/zed-industries/zed/assets/51882760/96884f36-5e88-400e-9a83-98435159e6c4


What I'm expecting:

https://github.com/zed-industries/zed/assets/51882760/c87d29a5-36d1-4621-8548-551173bc4a02

**Comment by alexkotusenko** [comment]
just add the following line to your settings if you want to disable it: `"ensure_final_newline_on_save": false,`

**Comment by Moshyfawn** [comment]
I believe the OP is aware of the `ensure_final_newline_on_save` setting, and the point of this issue is to determine whether an extra terminating newline makes sense in the case of an empty file, which doesn't seem to me to be the case, since at least on most Linux systems only a single line is enforced.

**Comment by deeptechdig** [comment]
zed 0.162.3, 
`"ensure_final_newline_on_save": false,`
this is not working for me on a remote ssh session. 
I added this line both in the global section and the host section. A new line is always added.

**Comment by injust** [comment]
My two cents: This is a terrible default. When have you ever needed to save a file with exactly one newline in it?

**Comment by esamattis** [comment]
Somewhat related, the `ensure_final_newline_on_save: true` also removes any extraneous lines from the end of the file. I would not expect it to ever remove lines just add the ending line if it is missing. This causes some unexpected diffs in my git commits.

**Comment by gab-txt** [comment]
I agree with @deeptechdig and @Moshyfawn: the use case we are talking about is having `ensure_final_newline_on_save` `true` while being able to save empty files without creating an additional empty row.

In my experience, this is the normal behavior when I use other editors: Zed's behavior is unexpected and, as far as I understand, the only solution is to switch to another editor to fix the double newlines.

**Comment by mstarodub** [comment]
`ensure_final_newline_on_save: true` is utterly broken.

- the issue in the original post can be reproduced by inserting some spaces, which will be trimmed, creating a file with a single `\n` (resulting in line 1 and 2) - saving an empty file without inserting spaces seems to be behaving correctly for me

- in addition to this, typing non-space characters into a new file, then saving it, does not result in a file with a final newline.
to reproduce: `cmd + n`-> type `aaa` -> `cmd + s` -> observe the final file does not have a newline at the end.

**Comment by MVHINC** [comment]
I have just tried "ensure_final_newline_on_save": false and restarted zed.
I still get a new line.
Zed 0.190.6

**Comment by michaelhashimoto** [comment]
I found a workaround, if you have both of these it works:

```
"ensure_final_newline_on_save": false,
"formatter": {
	"external": {
		"command": "sed",
		"arguments": ["-e", "s/ *$//"]
	}
}
```

There must be something in the default [formatter](https://zed.dev/docs/configuring-zed#formatter) that overrides this configuration.

I generally don't want trailing spaces anyways, so this configuration worked for me!

**Comment by IronicUsername** [comment]
thx @michaelhashimoto , it worked for me but i actually like the reverse :)

and for any traveler searching for a quick newline at the end of the file:
```
  "ensure_final_newline_on_save": true,
  "formatter": {
      "external": {
        "command": "sed",
        "arguments": ["-e", "s/[ \t]*$//", "-e", "$a\\"]
      }
    }
```

**Comment by CarlosUriass** [comment]
Could not reproduce on macOS Stable; creating and saving a new empty file does not add additional newlines.

Environment

OS: macOS 26.5.1 (arm64)
Zed: 1.14.2+stable.337.02abf5b08fa12c1c20a155ae3f796ef4c6c1a01e (Stable)

Steps

1. Created a new empty file (vim.txt).
2. Did not type anything into the file.
3. Saved the file.
4. Repeated the test with Vim mode enabled.
5. Checked the resulting file from the terminal using xxd vim.txt and wc -c vim.txt.

Result

Not reproducible on my end.

The file remains completely empty after saving. xxd vim.txt produces no output, and wc -c vim.txt reports 0 bytes, confirming that no newline characters are being inserted.

The same behavior was observed both with Vim mode disabled and enabled.

@yeskunall @AidanV

**Comment by injust** [comment]
I can reproduce with a file named `test.sh` on:

Zed: v1.14.2+stable.337.02abf5b08fa12c1c20a155ae3f796ef4c6c1a01e (Zed) 
OS: macOS 15.7.7
Memory: 16 GiB
Architecture: x86_64
